src/Processors/Formats/Impl/Parquet/RowGroupChunkReader.cpp

```diff
@@ -0,0 +1,508 @@
+            // intermediate column is already casted
+            if (removeNullableOrLowCardinalityNullable(src_col.column)->getDataType()
+                == removeNullableOrLowCardinalityNullable(types.at(i).type)->getColumnType())
+            {
+                casted_columns.emplace_back(std::move(src_col.column));
+            }
+            else
+            {
+                auto casted = castColumn(src_col, types.at(i).type);
+                casted_columns.emplace_back(casted);
+            }
+        }
+        return Chunk(std::move(casted_columns), rows_read);
+    }
+    else
+        return {};
+}
+
+arrow::io::ReadRange getColumnRange(const parquet::ColumnChunkMetaData & column_metadata)
+{
+    // From ComputeColumnChunkRange() in contrib/arrow/cpp/src/parquet/file_reader.cc:
+    //  > The Parquet MR writer had a bug in 1.2.8 and below where it didn't include the
+    //  > dictionary page header size in total_compressed_size and total_uncompressed_size
+    //  > (see IMPALA-694). We add padding to compensate.
+    int64_t col_start = column_metadata.data_page_offset();
+    if (column_metadata.has_dictionary_page() && column_metadata.dictionary_page_offset() > 0
+        && col_start > column_metadata.dictionary_page_offset())
+    {
+        col_start = column_metadata.dictionary_page_offset();
+    }
+    int64_t len = column_metadata.total_compressed_size();
+    return {col_start, len};
+}
+
+
+RowGroupPrefetch::RowGroupPrefetch(
+    SeekableReadBuffer & file_, std::mutex & mutex, const parquet::ArrowReaderProperties & arrow_properties_, ThreadPool & io_pool)
+    : file(file_), file_mutex(mutex), arrow_properties(arrow_properties_)
+{
+    callback_runner = threadPoolCallbackRunnerUnsafe<ColumnChunkData>(io_pool, "ParquetRead");
+}
+void RowGroupPrefetch::prefetchRange(const arrow::io::ReadRange & range)
+{
+    if (fetched)
+        throw Exception(ErrorCodes::PARQUET_EXCEPTION, "RowGroupPrefetch: prefetchColumnChunk called after startPrefetch");
+    ranges.emplace_back(range);
+}
+void RowGroupPrefetch::startPrefetch()
+{
+    if (fetched)
+        return;
+    fetched = true;
+    auto status = arrow::io::internal::CoalesceReadRanges(
+        ranges, arrow_properties.cache_options().hole_size_limit, arrow_properties.cache_options().range_size_limit);
+    if (!status.ok())
+    {
+        throw Exception(ErrorCodes::PARQUET_EXCEPTION, "Failed to coalesce read ranges: {}", status.status().message());
+    }
+    else
+        ranges = status.MoveValueUnsafe();
+    read_range_buffers.resize(ranges.size());
+    for (size_t i = 0; i < ranges.size(); i++)
+    {
+        auto & range = ranges[i];
+        read_range_buffers[i].range = range;
+        auto task = [this, range, i]() -> ColumnChunkData
+        {
+            auto & buffer = read_range_buffers[i].buffer;
+
+            buffer.resize(range.length);
+            int64_t count = 0;
+            if (file.supportsReadAt())
```

Comment:
What prevents use-after-free here?

Reply:
Didn't understand the question

Reply:
I thought `file` can be destroyed (as part of `ParquetBlockInputFormat`, e.g. if query fails with exception) while this `task` is executing/scheduled on the thread pool. But not I see that it's scheduled on `ParquetBlockInputFormat::io_pool`, and `~ParquetBlockInputFormat()` does `io_pool->wait()`, so it seems correct. Consider adding a comment about this, in case someone else gets confused the same way :)